### PR TITLE
Use Gemfile instead of core_gems.rb and plugin_gems.rb

### PR DESCRIPTION
**Pull request by kenhys**
NOTE: we need to wait activesupport 6.1 release to pass tests on Travis-CI.
(It conflict with td-agent requirements to tzinfo 2.0.2)

Closes: #126

### CODE REVIEW

**Comment by cosmo0920** [comment]
GitHub Actions complains that bundler is not found:
https://github.com/fluent-plugins-nursery/td-agent-builder/pull/142/checks?check_run_id=867723674#step:4:18

**Comment by cosmo0920** [comment]
This PR still causes the following error:

https://github.com/fluent-plugins-nursery/td-agent-builder/runs/876077221#step:4:17499
```
Circular dependency detected: TOP => build:all => build:licenses => build:gems => build:ruby_gems => build:fluentd => build:ruby_gems
/usr/share/gems/gems/rake-12.3.2/exe/rake:27:in `<top (required)>'
Tasks: TOP => build:all => build:licenses => build:gems => build:ruby_gems => build:fluentd
(See full trace by running task with --trace)
rake aborted!
```

**Comment by kenhys** [comment]
Before
plugin_gems->dl plugin/fluentd -> dl:flutentd/core->dl:core/ruby
After
ruby_gems->dl gems/fluentd -> dl/fluentd/ruby
Then circular dependency will be fixed.

**Comment by kenhys** [comment]
rake clean fails to run:

```
# rake clean
rake aborted!
Failed to download gem files!
/root/td-agent-4.0.0/td-agent/Rakefile:271:in `define_gem_files'
/root/td-agent-4.0.0/td-agent/Rakefile:156:in `define'
/root/td-agent-4.0.0/td-agent/Rakefile:1007:in `<top (required)>'
(See full trace by running task with --trace)
```

**Comment by kenhys** [comment]
It is caused by bundler incompatible version.

```
# bundle package --no-install --cache-path=downloads/
Don't run Bundler as root. Bundler can ask for sudo if it is needed, and installing your bundle as root will break this application for all non-root users on
this machine.
Fetching gem metadata from https://rubygems.org/............
Fetching gem metadata from https://rubygems.org/.
Resolving dependencies...
Bundler could not find compatible versions for gem "bundler":
  In Gemfile:
    bundler (= 2.1.4)

  Current Bundler version:
    bundler (1.16.1)
This Gemfile requires a different version of Bundler.
Perhaps you need to update Bundler by running `gem install bundler`?

Could not find gem 'bundler (= 2.1.4)' in any of the relevant sources:
  the local ruby installation
```

**Comment by kenhys** [comment]
We can install bundler explicitly.

`gem install bundler`

**Comment by kenhys** [comment]
`/usr/local/bin/bundle cache --no-install --cache-path=downloads/` fails.

It hits bundler bug?

```
nfortunately, an unexpected error occurred, and Bundler cannot continue.

First, try this link to see if there are any existing issue reports for this error:
https://github.com/bundler/bundler/search?q=same+file++%2Froot%2Ftd-agent-4.0.0%2Ftd-agent%2Fdownloads%2F%2Frake-13.0.1.gem+and+%2Froot%2Ftd-agent-4.0.0%2Ftd-agent%2Fdownloads%2Frake-13.0.1.gem&type=Issues

If there aren't any reports for this error yet, please create copy and paste the report template above into a new issue. Don't forget to anonymize any private data! The new issue form is located at:
https://github.com/bundler/bundler/issues/new
```

## Backtrace

```
ArgumentError: same file: /root/td-agent-4.0.0/td-agent/downloads//rake-13.0.1.gem and /root/td-agent-4.0.0/td-agent/downloads/rake-13.0.1.gem
  /var/lib/gems/2.5.0/gems/bundler-2.1.4/lib/bundler/vendor/fileutils/lib/fileutils.rb:1587:in `block in fu_each_src_dest'
  /var/lib/gems/2.5.0/gems/bundler-2.1.4/lib/bundler/vendor/fileutils/lib/fileutils.rb:1602:in `fu_each_src_dest0'
  /var/lib/gems/2.5.0/gems/bundler-2.1.4/lib/bundler/vendor/fileutils/lib/fileutils.rb:1586:in `fu_each_src_dest'
  /var/lib/gems/2.5.0/gems/bundler-2.1.4/lib/bundler/vendor/fileutils/lib/fileutils.rb:429:in `cp'
  /var/lib/gems/2.5.0/gems/bundler-2.1.4/lib/bundler/source/rubygems.rb:206:in `cache'
  /var/lib/gems/2.5.0/gems/bundler-2.1.4/lib/bundler/runtime.rb:132:in `block in cache'
  /var/lib/gems/2.5.0/gems/bundler-2.1.4/lib/bundler/spec_set.rb:147:in `each'
  /var/lib/gems/2.5.0/gems/bundler-2.1.4/lib/bundler/spec_set.rb:147:in `each'
  /var/lib/gems/2.5.0/gems/bundler-2.1.4/lib/bundler/runtime.rb:128:in `cache'
  /var/lib/gems/2.5.0/gems/bundler-2.1.4/lib/bundler/cli/install.rb:67:in `run'
  /var/lib/gems/2.5.0/gems/bundler-2.1.4/lib/bundler/cli/cache.rb:33:in `install'
  /var/lib/gems/2.5.0/gems/bundler-2.1.4/lib/bundler/cli/cache.rb:17:in `run'
  /var/lib/gems/2.5.0/gems/bundler-2.1.4/lib/bundler/cli.rb:461:in `cache'
  /var/lib/gems/2.5.0/gems/bundler-2.1.4/lib/bundler/vendor/thor/lib/thor/command.rb:27:in `run'
  /var/lib/gems/2.5.0/gems/bundler-2.1.4/lib/bundler/vendor/thor/lib/thor/invocation.rb:127:in `invoke_command'
  /var/lib/gems/2.5.0/gems/bundler-2.1.4/lib/bundler/vendor/thor/lib/thor.rb:399:in `dispatch'
  /var/lib/gems/2.5.0/gems/bundler-2.1.4/lib/bundler/cli.rb:30:in `dispatch'
  /var/lib/gems/2.5.0/gems/bundler-2.1.4/lib/bundler/vendor/thor/lib/thor/base.rb:476:in `start'
  /var/lib/gems/2.5.0/gems/bundler-2.1.4/lib/bundler/cli.rb:24:in `start'
  /var/lib/gems/2.5.0/gems/bundler-2.1.4/exe/bundle:46:in `block in <top (required)>'
  /var/lib/gems/2.5.0/gems/bundler-2.1.4/lib/bundler/friendly_errors.rb:123:in `with_friendly_errors'
  /var/lib/gems/2.5.0/gems/bundler-2.1.4/exe/bundle:34:in `<top (required)>'
  /usr/local/bin/bundle:23:in `load'
  /usr/local/bin/bundle:23:in `<main>'
```

**Comment by kenhys** [comment]
Definitely no.

```
/usr/local/bin/bundle cache --no-install --cache-path=downloads/ fails.
```

should be `--cache-path=downloads`.

**Comment by kenhys** [comment]
I've squashed local commit,

**Comment by cosmo0920** [comment]
CI is still failing:

```log
dpkg-source: info: local changes detected, the modified files are:
 td-agent-4.0.0/td-agent/Gemfile.lock
```

ref: https://github.com/fluent-plugins-nursery/td-agent-builder/pull/142/checks?check_run_id=888358760#step:4:4913

**Comment by kenhys** [comment]
Only Travis buster and focal is failed.

**Comment by kenhys** [comment]
It is strange behavior.


```
/usr/bin/ruby2.5 -S rake clean

Don't run Bundler as root. Bundler can ask for sudo if it is needed, and

installing your bundle as root will break this application for all non-root

users on this machine.

rm -rf .bundle

rm -rf Gemfile.lock

fatal: not a git repository (or any of the parent directories): .git

cd -

dh_auto_clean

make[1]: Leaving directory '/build/td-agent-4.0.0'

   dh_clean

 dpkg-source -b .

dpkg-source: info: using source format '3.0 (quilt)'

dpkg-source: info: building td-agent using existing ./td-agent_4.0.0.orig.tar.gz

dpkg-source: warning: ignoring deletion of file td-agent/downloads/thread_safe-0.3.6.gem, use --include-removal to override

dpkg-source: warning: ignoring deletion of file td-agent/downloads/net-http-pipeline-1.0.1.gem, use --include-removal to override

dpkg-source: warning: ignoring deletion of file td-agent/downloads/gh-0.18.0.gem, use --include-removal to override

dpkg-source: warning: ignoring deletion of file td-agent/downloads/ethon-0.12.0.gem, use --include-removal to override

dpkg-source: warning: ignoring deletion of file td-agent/downloads/highline-2.0.3.gem, use --include-removal to override

dpkg-source: warning: ignoring deletion of file td-agent/downloads/i18n-1.8.4.gem, use --include-removal to override

dpkg-source: warning: ignoring deletion of file td-agent/downloads/launchy-2.4.3.gem, use --include-removal to override

dpkg-source: warning: ignoring deletion of file td-agent/downloads/json-2.3.1.gem, use --include-removal to override

dpkg-source: warning: ignoring deletion of file td-agent/downloads/websocket-1.2.8.gem, use --include-removal to override

dpkg-source: warning: ignoring deletion of file td-agent/downloads/tzinfo-1.2.7.gem, use --include-removal to override

dpkg-source: warning: ignoring deletion of file td-agent/downloads/net-http-persistent-2.9.4.gem, use --include-removal to override

dpkg-source: warning: ignoring deletion of file td-agent/downloads/faraday_middleware-1.0.0.gem, use --include-removal to override

dpkg-source: warning: ignoring deletion of file td-agent/downloads/typhoeus-0.8.0.gem, use --include-removal to override

dpkg-source: warning: ignoring deletion of file td-agent/downloads/travis-1.9.1.gem, use --include-removal to override

dpkg-source: warning: ignoring deletion of file td-agent/downloads/minitest-5.14.1.gem, use --include-removal to override

dpkg-source: warning: ignoring deletion of file td-agent/downloads/pusher-client-0.6.2.gem, use --include-removal to override

dpkg-source: warning: ignoring deletion of file td-agent/downloads/activesupport-5.2.4.3.gem, use --include-removal to override

dpkg-source: error: cannot represent change to td-agent/downloads/async-1.26.2.gem: binary file contents changed

dpkg-source: error: add td-agent/downloads/async-1.26.2.gem in debian/source/include-binaries if you want to store the modified binary in the debian tarball
```

**Comment by kenhys** [comment]
It seems that travis_retry has nothing to do with it. https://github.com/fluent-plugins-nursery/td-agent-builder/pull/149/commits/8ac528e9e74d0a5b01e7234c946647ea7b627ada
https://travis-ci.com/github/fluent-plugins-nursery/td-agent-builder/jobs/362736715

**Comment by kenhys** [comment]
https://travis-ci.com/github/fluent-plugins-nursery/td-agent-builder/jobs/362736715#L1004 gem is copied into tar.gz

**Comment by kenhys** [comment]
With local build for buster, it seems ok (some gem was updated) (diff with td-agent 4.0.0-1.deb)
 [local-build.buster.diff.txt](https://github.com/fluent-plugins-nursery/td-agent-builder/files/4950594/local-build.buster.diff.txt)

**Comment by kenhys** [comment]
deian/source/include-binaries is a workaround, but I think that it is not suitable solution.

**Comment by ashie** [comment]
When I run `rake:apt` locally, I get the following error:

```
$ rake apt:build
cd td-agent
/usr/bin/ruby2.7 -S rake apt:build
ERROR:  While executing gem ... (Gem::FilePermissionError)
    You don't have write permissions for the /var/lib/gems/2.7.0 directory.
Following files may not be writable, so sudo is needed:
  /usr/local/bin
  /var/lib/gems/2.7.0
  /var/lib/gems/2.7.0/build_info
  /var/lib/gems/2.7.0/cache
  /var/lib/gems/2.7.0/doc
  /var/lib/gems/2.7.0/extensions
  /var/lib/gems/2.7.0/gems
  /var/lib/gems/2.7.0/specifications


Your user account isn't allowed to install to the system RubyGems.
  You can cancel this installation and run:

      bundle install --path vendor/bundle

  to install the gems into ./vendor/bundle/, or you can enter your password
  and install the bundled gems to RubyGems using sudo.
```

**Comment by kenhys** [comment]
It could be reproducible with the following command on local PC.

* bundle install
* rake apt:build

**Comment by kenhys** [comment]
> When I run `rake:apt` locally, I get the following error:
> 
> ```
> $ rake apt:build
> cd td-agent
> /usr/bin/ruby2.7 -S rake apt:build
> ERROR:  While executing gem ... (Gem::FilePermissionError)
>     You don't have write permissions for the /var/lib/gems/2.7.0 directory.

Ah, I've got it.

**Comment by kenhys** [comment]
rvm system reproduce it.(While executing gem ... (Gem::FilePermissionError))

**Comment by kenhys** [comment]
> これって手で作成しています？それともスクリプトで作成しています？
手で作成しているとするとメンテナンスが大変なのでrakeタスクにするとどうでしょうか @kenhys?

↑は賛成なんですが、ちょっとややこしいんですよね。
どうもTOP_DIR/Gemfileのbundle installでtd-agent/downloadsにgemがダウンロードされ、[1]
rake buildでTOP_DIR/td-agent/Gemfileに記述したgemがtd-agent/downloads/にgemがダウンロードされる [2]
[1]のgemは[2]のときに削除される。
td-agent/debian/source/include-binariesには[1]だけでなく[2]もリストアップしないといけないみたいです。[3]
ちょっと不可解な挙動なんですよね。[3]が。

**Comment by cosmo0920** [comment]
> ↑は賛成なんですが、ちょっとややこしいんですよね。
> どうもTOP_DIR/Gemfileのbundle installでtd-agent/downloadsにgemがダウンロードされ、[1]
> rake buildでTOP_DIR/td-agent/Gemfileに記述したgemがtd-agent/downloads/にgemがダウンロードされる [2]
> [1]のgemは[2]のときに削除される。
> td-agent/debian/source/include-binariesには[1]だけでなく[2]もリストアップしないといけないみたいです。[3]
> ちょっと不可解な挙動なんですよね。[3]が。Travis-CIの場合なんですが。

むむ、ちょっとややこしいですね。。。

**Comment by kenhys** [comment]
> むむ、ちょっとややこしいですね。。。

ちょっとうまくやれないか考えてみます。

**Comment by kenhys** [comment]
Added task to generate debian/source/include-binries. Note that system's ruby should be used.

**Comment by kenhys** [comment]
Just ignore failure of Travis-CI untill activesupport 6.1 will be released.

File: Gemfile [hunk 1]
```diff
@@ -4,5 +4,67 @@ source "https://rubygems.org"
...
+  gem "windows-pr", "1.2.6"
+  gem "winevt_c", "0.8.1"
+else
+  # temporal solution for ruby 2.7
+  gem "nokogiri", "1.10.10"
+end
+
+if ENV["TRAVIS"]
+  # Travis CI fails because activesupport requires tzinfo ~> 1.1.
+  # We need to wait until activesupport 6.1.
```
Comment (cosmo0920) [inline Gemfile:67]: :+1:
We also should work for upgrading fluent-plugin-sql to depend on activesupport 6.1 or later.

**Comment by cosmo0920** [comment]
I'm confirming to build with this changes.

**Comment by kenhys** [comment]
match condition to RUBY_PLATFORM should be fixed with x64-mingw32.

**Comment by kenhys** [comment]
https://rubygems.org/gems/nokogiri/versions/1.10.10-x64-mingw32 should be downloaded by ` bundle package --no-install --cache-path=td-agent/downloads`, but it fails.

File: Gemfile [hunk 2]
```diff
@@ -49,6 +48,7 @@ if /mswin|mingw/ =~ RUBY_PLATFORM
   gem "ffi-win32-extensions", "1.0.3"
   gem "fluent-plugin-parser-winevt_xml", "0.2.2"
   gem "fluent-plugin-windows-eventlog", "0.7.0"
+  gem "rdkafka", "0.8.0"
```
Comment (cosmo0920) [inline Gemfile:51]: え、これelse節の中に入れないとWindowsでビルドされるのでは？

Comment (kenhys) [reply Gemfile]: あ、へんなことやっていますね。。。直します。

**Comment by kenhys** [comment]
Need to care gem handling a bit.

**Comment by cosmo0920** [comment]
Windows Installer can be created correctly. But still problem exists.
The issue is overwriting rake executable on Windows due to conflicting RubyInstaller bundled rake.

**Comment by kenhys** [comment]
Blocker about this PR is generated diff in .bundle/config( BUNDLE_CACHE_PATH)

**Comment by kenhys** [comment]
* `bundle package --no-install --cache-path=#{DOWNLOADS_DIR}` in define_gem_files rewrite .bundle/config.

**Comment by kenhys** [comment]
`bundle package --no-install --cache-path=#{DOWNLOADS_DIR}` in define_gem_files in command line execution doesn't affects them.

**Comment by kenhys** [comment]
Bundler.Settings.set_command_option_if_given stores full path?

**Comment by kenhys** [comment]
It seems around https://github.com/ruby/ruby/blob/master/lib/bundler/settings.rb#L103?

**Comment by kenhys** [comment]
Bundler.Settings.set_command_option_if_given accepts full path.

**Comment by kenhys** [comment]
log prints `Running `bundle cache --cache-path "/work/fluentd/td-agent-builder.devel/td-agent/downloads" --no-install` with bundler 2.1.4`, thus CLI::Cache.new(options) already accepts full path.

**Comment by kenhys** [comment]
https://github.com/ruby/ruby/blob/master/lib/bundler/cli/install.rb#L167
It seems no option to keep relative path.

**Comment by kenhys** [comment]
@cosmo0920 are there still any blocker to merge?

**Comment by cosmo0920** [comment]
I just got merged. Thank you!

**Pull request merged**